apache/pinot
# [Backfill] flink based upsert table backfill support

`feature` `release-notes` `backward-incompatible`

Follow up of #13107

Delivers: https://github.com/apache/pinot/issues/12987, https://github.com/apache/pinot/issues/10896

This PR adds support for uploading batch generated segments to an upsert table. A new segmentName convention `UploadedRealtimeSegmentName` was introduced which enables segment assignment for uploaded segments(not generated by stream consumers) based on the provided partition number.

Changes:
- Fixes an edge case bug with segment flush happening on flushThreshold + 1
- Adds new config in BatchConfigMap needed for `UploadedRealtimeSegmentNameGenerator`
- Flink PinotSinkFunction uses `UploadedRealtimeSegmentNameGenerator` for generating segments for Upsert tables for correct segment assignment.
- Adds support to configure segment name prefix in Flink PinotSinkFunction
- Adds support to configure segment creation time for segments generated as part of a run of Flink workload. This is helpful to group segments of a run with the same date and if needed refresh/overwrite using a rerun.
- Adds a new integration test for upsert tables
- Refactor README with an example to quick start.

Tagged backward-incompatible as it changes the constructor signature for FlinkSegmentWriter

Test plan:
- Integration test
    - Test partitioned parallel segment creation and uplaod
    - Test partitioned parallel segment creation with increasing sequenceId
- Cluster test:
  - Set up a new upsert table with keys partitioned using murmur2
  - Generate segments from a Flink job and upload to table
  - Validation:
    - Check the keys of a uploaded segment are consistently partitioned with realtime stream generated segments
    - <img width="760" alt="Screenshot 2024-08-19 at 2 26 49 PM" src="https://github.com/user-attachments/assets/56466699-cfc9-4155-b0e1-07ac04fcc02e">
